### DISCUSSION TITLE
Link_pred.py  using GAE produce 3M+ edge index as the decoded edge index size!

### DISCUSSION THREAD

**Discussion by afshinebtia** [pyg-team/pytorch_geometric#7212] [Q&A] [answered]
Hi,
In the "Link_pred.py " in the example folder which uses GAE for link prediction in Cora dataset, at the end of the code, when we decode "Z", the size of the decoded edge_index is 3M+ !!! However, the Cora has 10556 edges!

Thanks in advance

**Comment by rusty1s** [comment, marked as answer]
Yes, it produces a score for every possible node pair. With `2708` nodes, this amounts to `2708*2708/2 = 3.6M` pairs (divided by half since the graph is undirected).

**Reply by afshinebtia** [reply]
Thanks for your response.
According to one of your replies to use the following command to find the test_probability;

z = model.encode(test_data.x, test_data.edge_index)
test_prob = (z.unsqueeze(0) * z.unsqueeze(1)).sum(dim=-1)

Can I pass the test_prob through a "softmax" to make the test_prob values between 0 and 1, then select the highest probability, e.g., above 0.90 as the connected edges?

**Reply by rusty1s** [reply]
Yeah, that is possible, but you would rather use `sigmoid` here instead of `softmax`. Alternatively, I suggest to use something like `torch.topk` to find the top-k edges for each node.

**Comment by afshinebtia** [comment]
Yeah, you're completely right.
Thanks for your response.

**Reply by afshinebtia** [reply]
I also have another question;

In the transductive link prediction task, during the training stage, all the nodes are visible to the model; however, only some parts (let's say 90% for training) of the negative and positive edges are being used for training purposes. Now my question is can we have isolated nodes in our graph and use the trained model to predict the possible links between that isolated node and the other nodes in the graph?

**Reply by rusty1s** [reply]
Yes, this is possible, but it is hard to tell whether this gives good model performance since the node embeddings obtained from the GNN will be limited to the feature set of the isolated node.

**Comment by afshinebtia** [comment]
Hi again Matthias,

I have an issue creating a Data object in PyG which contains a list of graphs to use this GAE link prediction in an inductive way on those sets of graphs (multi-graphs). I have an adj_list and node_features list, and defined my data_list in PyG 2.xx as: 

data_list = [ ]
for i in range(len(adj_list)):
    adj = torch.tensor(adj_list[i], dtype=torch.float)
    node_features = torch.tensor(node_features_list[i], dtype=torch.float)
    node_labels = torch.tensor(node_label_list[i], dtype=torch.float)
    edge_index = torch.tensor(adj.nonzero(), dtype=torch.long).t()
    data = Data(x=node_features, edge_index=edge_index, y =  node_labels)
    data_list.append(data)

in the end, when I print the len(data_list) it gives me the correct number of graphs that I have. Also, for a single graph, e.g., data_list[0], it gives me the correct shape for "x", "y", and "edge_index" as: 

The 0th graph in data_list has:  Data(x=[33, 2], edge_index=[2, 69], y=[33, 1]

Now my problem is although in MUTAG / TUDataset (https://colab.research.google.com/drive/1I8a0DfQ3fI7Njc62__mVXUlcAleUclnb?usp=sharing#scrollTo=mHSP6-RBOqCE), it is possible to access "dataset.num_features" directly, in my data_list, when I run "data_list.num_features", I get the following error:  " AttributeError: 'list' object has no attribute 'num_node_features' "

Also, when I use "data_list_pyg = torch_geometric.data.Data(data_list)", the error is solved, but the "len" of the "data_list_pyg" is shown as 1 and the number of features is shown as 0, however, it should be 91 and 2. respectively.


Would you please elaborate more that how can I have the same format of multi-graph datasets in PyG such as MUTAG to be able to get "data_list.num_features" without any error?


Thanks

**Reply by rusty1s** [reply]
Since `data_list` is a Python list, it does not support the `InMemoryDataset` property `num_features`. A quick workaround is to call `data_list[0].num_features` instead or just hard-code it to `2`.

**Reply by afshinebtia** [reply]
Thanks.
Do you have any examples of "In Memory Dataset" in PyG? I read the manual sheet of PyG, but it's not clear for me.

**Reply by rusty1s** [reply]
All our datasets in `torch_geometric.datasets` are implementations of `InMemoryDataset`, which is basically just a wrapper around holding a list of Python objects.

**Reply by afshinebtia** [reply]
Great!
Thanks for your response.
